mail: allow user to redirect FQDN hostname to imprint

mkg20001 opened the pull request:
fixes PL-129598

@flyingcircusio/release-managers

## Release process

Impact:

Changelog:
- adds imprintUrl and imprintText to allow the user to specify either a bunch of HTML or a link to redirect the FQDN to so that t-online has an imprint to look at

## Security implications

- [x] [Security requirements](https://wiki.flyingcircus.io/System_Development_Guideline#Security_requirement_principles_and_testing) defined? (WHERE)
  - imprint is served over tls
- [x] Security requirements tested? (EVIDENCE)
  - addSSL is set, tested on testvm

ctheune commented on nixos/roles/mailserver.nix:
```diff
@@ -102,6 +102,24 @@ in
+      imprintText = mkOption {
+        type = with types; nullOr str;
+        description = ''
+          Imprint as HTML text to be in compilance with T-Online Postmaster Rules
```
Verweis/Link auf postmaster.t-online.de und noch zum TMG wo die Anforderungen an die  Anbieterkennzeichnung beschrieben sind.

mkg20001 replied:
Hat mir wer ein guten link den ich da verlinken soll zum TMG?

dpausp replied:
https://www.gesetze-im-internet.de/tmg/__5.html ?

ctheune commented on nixos/roles/mailserver.nix:
```diff
@@ -102,6 +102,24 @@ in
         '';
       };
 
+      imprintUrl = mkOption {
```
Wir sollten das zur Mindestanforderung machen: es muss eins gesetzt sein (aber es dürfen nicht beide gesetzt sein).

mkg20001 replied:
```
error: 
Failed assertions:
- The imprintUrl or imprintText option needs to be set

This is mandated by:
- Telekom Postmaster Rules §4.1 (see https://postmaster.t-online.de/#t4.1), for sending mail to Telekom customers
- Telemediengesetz Anbieterkennzeichnung (see #todo#), for compilance with German law

(use '--show-trace' to show detailed location information)
ERROR: Building channel failed!
```

dpausp commented on nixos/services/mail/default.nix:
```diff
@@ -86,6 +86,21 @@ in {
     let
       fqdn = with config.networking; "${hostName}.${domain}";
     in {
+      warnings = optional (role.imprintUrl != null || role.imprintText != null) ''
```
optional is not defined here

dpausp commented on nixos/services/mail/default.nix:
```diff
@@ -86,6 +86,21 @@ in {
+        Please do it as soon as possible
+      '';
+      assertions = [
+        { assertion = role.imprintUrl != null && role.imprintText != null;
```
the assertion must be turned around: at the moment it fails when none or one of the options is set.

dpausp commented on nixos/services/mail/default.nix:
```diff
@@ -200,7 +215,27 @@ in {
           } config.fclib.network.fe.dualstack.addressesQuoted);
-        in listToAttrs (map cfgForDomain role.domains);
+        in listToAttrs ((map cfgForDomain role.domains) ++
+          [ (lib.nameValuePair fqdn
```
This must point to the `role.mailHost`, not `fqdn` here. (it's a bit misleading in the code because `mailserver.fqdn = role.mailHost` but there's a `fqdn` in a let that points to the internal hostname like test33.fcio.net).

After fixing this, there will be a problem with nixos/services/mail/roundcube.nix, line 51 which redirects `role.mailHost` to the webmailer. This has to be removed. Also, we must assert that `role.webmailHost != role.mailHost` to avoid a conflict between imprint and webmailer...

Still an open question: we already have some servers that use webmailHost == mailHost, what to do here?

dpausp replied:
New plan: keep the redirect to webmailHost as platform default, but it can be overridden by setting imprintText or imprintUrl. Both options should be optional and not generate a warning. If mailHost == webmailHost, both options should produce an error.
We have to document this properly in the mailserver role docs.

dpausp commented on nixos/services/mail/default.nix:
```diff
@@ -200,7 +215,27 @@ in {
+              (lib.mkIf (role.imprintUrl == null && role.imprintText == null) {
+                root = with pkgs; writeTextDir "index.html" ''
+                  <h1>Imprint missing</h1>
+                  <h2>Imprint can be configured using <tt>flyingcircus.roles.mailserver.imprintText</tt> or <tt>flyingcircus.roles.mailserver.imprintUrl</tt></h2>
```
A link to the documentation at https://doc.flyingcircus.io/roles/fc-21.05-production/mailserver.html#how-do-i-perform-a-basic-setup would be helpful.

dpausp commented on nixos/services/mail/default.nix:
```diff
@@ -86,6 +86,21 @@ in {
+        The imprintUrl or imprintText option needs to be set
+
+        This is mandated by:
+        - Telekom Postmaster Rules §4.1 (see https://postmaster.t-online.de/#t4.1), for sending mail to Telekom customers
```
This should in the documentation, too.

dpausp reviewed (changes requested):
See comments. Also needs a rebase, I changed the documentation in the meantime.

dpausp commented on nixos/services/mail/default.nix:
```diff
@@ -104,6 +104,21 @@ in {
       fqdn = with config.networking; "${hostName}.${domain}";
     in {
+      assertions = [
+        { assertion = !(role.imprintUrl != null && role.imprintText != null);
```
The double negation is hard to read, this looks better with an `||`

dpausp commented on doc/src/mailserver.rst:
```diff
@@ -48,6 +48,12 @@ which will be advertised as MX name on your mail domain. This host name (called
+
+For this you can either set imprintUrl to the location of your existing imprint,
+or use imprintText to specify an imprint in HTML format
+
```
Tell the users about the webmailHost behaviour: what's the default and what will lead to an error (assertions)?

dpausp commented on doc/src/mailserver.rst:
```diff
@@ -48,6 +48,12 @@ which will be advertised as MX name on your mail domain. This host name (called
 and reverse lookups.
 
+Additionally, some mail providers (namely [Telekom/T-Online](https://postmaster.t-online.de/#t4.1))
+require that your mailserver has an imprint served at it's hostname.
```
we are not really sure what they want, I'd say `may require`

`it's` -> `its`

dpausp commented on nixos/services/mail/default.nix:
```diff
@@ -104,6 +104,21 @@ in {
+            Remove one to fix this issue
+          '';
+        }
+        { assertion = !((role.imprintUrl != null && role.imprintText != null) && (role.mailHost == role.webmailHost));
```
The first part is true if both options are set which isn't possible. Avoid double negation here as well.

dpausp commented on nixos/services/mail/default.nix:
```diff
@@ -104,6 +104,21 @@ in {
+        }
+        { assertion = !((role.imprintUrl != null && role.imprintText != null) && (role.mailHost == role.webmailHost));
+          message = ''
+            imprintUrl/imprintText cannot be set when webmail ${role.webmailHost} is already being served under ${role.mailHost}
```
I'd remove the `${role.webmailHost}` because it's the same as `${role.mailHost}` which looks strange in the final output.

dpausp commented on nixos/services/mail/default.nix:
```diff
@@ -218,7 +233,27 @@ in {
+        in listToAttrs ((map cfgForDomain (attrNames (filterAttrs (domain: config: config.enable && config.autoconfig) role.domains))) ++
+          (optional (role.imprintUrl != null || role.imprintText != null ) (lib.nameValuePair role.mailHost
+            (lib.mkMerge [
+              {
```
listen addresses must be set the the frontend IPs here or Nginx will prefer the autoconfig vhosts (they are already listening on fe only)

mkg20001 commented:
Following problem:

module system gives us the values without priorities, just the clean, resolved values

we need the values with their priorities so when doing virtualHosts = fc.virtualHosts the module system knows which were defaults and which were user-set values from fc.virtualHosts

I have no idea how to do that

Alternative is to use options.flyingcircus.services.nginx.virtualHosts.definitions (raw user-definied keys), use modules.mergeDefinitions and go through each of the nameValuePairs, applying the FC changes like with the JSON previously.

also noticed other stuff:
- currently listenAddress and listenAddress6 default to 0.0.0.0 and [::] respectivly, meaning that the default that should use frontend addresses for listenAddresses will never trigger. I think that's wrong and all fc.virtualHosts should use FE addresses by default - correct?

mkg20001 committed:
mail: optional imprintText/imprintUrl option

This option allows the user to set an imprintText/imprintUrl which the
virtualHost under $mailHost will display or redirect to

This fixes PL-129598

Also:
- update documentation and mention t-online imprint requirement
- move roundcube webmail redirect to mailserver role
- add assertion that imprintText/imprintUrl cannot be used if webmail
already served under $mailHost
- document possible imprint requirement

mkg20001 committed:
nginx: make frontend default for non-json

mkg20001 committed:
mail: use new domains format in test
